Asterisk.2013/Asterisk.NET/Manager/ManagerConnection.cs

```diff
@@ -15,1566 +15,919 @@
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(DTMFEvent), arg => fireEvent(DTMF, arg));
+
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(VarSetEvent), arg => fireEvent(VarSet, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(AGIExecEvent), arg => fireEvent(AGIExec, arg));
+
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(ConfbridgeStartEvent), arg => fireEvent(ConfbridgeStart, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(ConfbridgeJoinEvent), arg => fireEvent(ConfbridgeJoin, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(ConfbridgeLeaveEvent), arg => fireEvent(ConfbridgeLeave, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(ConfbridgeEndEvent), arg => fireEvent(ConfbridgeEnd, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(ConfbridgeTalkingEvent), arg => fireEvent(ConfbridgeTalking, arg));
+
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(FailedACLEvent), arg => fireEvent(FailedACL, arg));
+
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(AttendedTransferEvent), arg => fireEvent(AttendedTransfer, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(BridgeCreateEvent), arg => fireEvent(BridgeCreate, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(BridgeDestroyEvent), arg => fireEvent(BridgeDestroy, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(BridgeEnterEvent), arg => fireEvent(BridgeEnter, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(BridgeLeaveEvent), arg => fireEvent(BridgeLeave, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(BlindTransferEvent), arg => fireEvent(BlindTransfer, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(DialBeginEvent), arg => fireEvent(DialBegin, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(DialEndEvent), arg => fireEvent(DialEnd, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(QueueCallerJoinEvent), arg => fireEvent(QueueCallerJoin, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(QueueCallerLeaveEvent), arg => fireEvent(QueueCallerLeave, arg));
+            Helper.RegisterEventHandler(registeredEventHandlers, typeof(QueueMemberPauseEvent), arg => fireEvent(QueueMemberPause, arg));
 
             #endregion
 
-            this.internalEvent += new ManagerEventHandler(internalEventHandler);
-		}
-		#endregion
-
-		#region Constructor - ManagerConnection(hostname, port, username, password)
-		/// <summary>
-		/// Creates a new instance with the given connection parameters.
-
```

Comment:
if i add <code>return</code> after this line, then the condition <code>if (fireAllEvents)</code> is never will called.

Reply:
Ooh yeah... But in that case you changed something else to make [`FireAllEvents`](http://asternet.github.io/AsterNET/html/P_AsterNET_Manager_ManagerConnection_FireAllEvents.htm) not work as intended.
Can't look at the mega diff on my phone, so I'll look it over tomorrow.

Reply:
In the current code, around [line 1284](https://github.com/AsterNET/AsterNET/blob/26c41504add0fc69a283da37a80640a02e1ca0ad/Asterisk.2013/Asterisk.NET/Manager/ManagerConnection.cs#L1284) and such.
The switch case assures that only known events are handled. If an event is unknown to AsterNET it will fire an `UnhandledEvent` as a way of saying "unknown event?!".
The reason you have to return after either result is that `fireAllEvents` then allows you to see all events that you aren't currently subscribed it get fired as `UnhandledEvent` instead.

```
int eventHash = e.GetType().Name.GetHashCode();
if (registeredEventHandlers.ContainsKey(eventHash))
{
	switch (registeredEventHandlers[eventHash])
	{
		case 0:
			if (AgentCallbackLogin != null)
			{
				AgentCallbackLogin(this, (AgentCallbackLoginEvent)e);
				return;
			}
			break;
		...
		default:
			if (UnhandledEvent != null)
				UnhandledEvent(this, e);
			return;
	}
}
if (fireAllEvents && UnhandledEvent != null)
	UnhandledEvent(this, e);
```

Reply:
Now we do not think about indexes. The hash table stores all calls. Changed the event call. Now, if there is no event in the hash table or the user is not subscribed to the events, then we call <code>if (fireAllEvents) { fireEvent(UnhandledEvent, e); }</code>